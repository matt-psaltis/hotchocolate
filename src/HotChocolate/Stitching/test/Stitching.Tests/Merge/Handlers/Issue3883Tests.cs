using System.Collections.Generic;
using System.Linq;
using HotChocolate.Language;
using HotChocolate.Language.Utilities;
using Snapshooter.Xunit;
using Xunit;

namespace HotChocolate.Stitching.Merge.Handlers
{
    public class Issue3883Tests
    {
        [Fact]
        public void Merge_SchemasWithConflictingComplexTypes_TypesMerge()
        {
            // arrange
            DocumentNode schema_a =
                Utf8GraphQLParser.Parse(@"
type CompanyConnection {
  pageInfo: PageInfo!
  edges: [CompanyEdge!]
  nodes: [Company]
}

type CompanyEdge {
  cursor: String!
  node: Company
}

type Company {
  id: String!
  name1: String
}");
            DocumentNode schema_b =
                Utf8GraphQLParser.Parse(@"
type CompanyConnection {
  pageInfo: PageInfo!
  edges: [CompanyEdge!]
  nodes: [Company]
}

type CompanyEdge {
  cursor: String!
  node: Company
}

type Company {
  id: String!
  name3: String
}");
            DocumentNode schema_c =
                Utf8GraphQLParser.Parse(@"
type CompanyConnection {
  pageInfo: PageInfo!
  edges: [CompanyEdge!]
  nodes: [Company]
}

type CompanyEdge {
  cursor: String!
  node: Company
}

type Company {
  id: String!
  name3: String
}");

            var types = BuildTypes("Schema_A", schema_a)
                .Concat(BuildTypes("Schema_B", schema_b))
                .Concat(BuildTypes("Schema_C", schema_c))
                .ToList();

            var context = new SchemaMergeContext();

            // act
            var typeMerger = new ObjectTypeMergeHandler((c, t) => { });
            typeMerger.Merge(context, types);

            // assert
            context
                .CreateSchema()
                .Print()
                .MatchSnapshot();
        }

        private static IEnumerable<ITypeInfo> BuildTypes(string schemaName, DocumentNode schema)
        {
            var schemaInfo = new SchemaInfo(schemaName, schema);
            return schema.Definitions.OfType<ITypeDefinitionNode>()
                .Select(x => TypeInfo.Create(x, schemaInfo));
        }
    }
}
