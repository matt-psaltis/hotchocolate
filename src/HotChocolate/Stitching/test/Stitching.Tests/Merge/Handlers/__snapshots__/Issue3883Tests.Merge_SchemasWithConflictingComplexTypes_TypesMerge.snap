type CompanyConnection @source(name: "CompanyConnection", schema: "Schema_A") {
  pageInfo: PageInfo!
  edges: [CompanyEdge!]
  nodes: [Company]
}

type CompanyEdge @source(name: "CompanyEdge", schema: "Schema_A") {
  cursor: String!
  node: Company
}

type Company @source(name: "Company", schema: "Schema_A") {
  id: String!
  name1: String
}

type Schema_B_CompanyConnection @source(name: "CompanyConnection", schema: "Schema_B") {
  pageInfo: PageInfo!
  edges: [CompanyEdge!]
  nodes: [Company]
}

type Schema_B_CompanyEdge @source(name: "CompanyEdge", schema: "Schema_B") {
  cursor: String!
  node: Company
}

type Schema_B_Company @source(name: "Company", schema: "Schema_B") {
  id: String!
  name3: String
}

type Schema_C_CompanyConnection @source(name: "CompanyConnection", schema: "Schema_C") {
  pageInfo: PageInfo!
  edges: [CompanyEdge!]
  nodes: [Company]
}

type Schema_C_CompanyEdge @source(name: "CompanyEdge", schema: "Schema_C") {
  cursor: String!
  node: Company
}

type Schema_C_Company @source(name: "Company", schema: "Schema_C") {
  id: String!
  name3: String
}
